# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: ipc/glue/UtilityProcessChild.cpp
# Commit: d56484304ec3
# Full Hash: d56484304ec3abe3891ede9e8d141c95543d3e37
# Author: Ray Kraesig <rkraesig@mozilla.com>
# Date: 2023-09-16 09:14:45
# Regressor Bug: 1837079
# File Overlap Count: 2
# Description:
#   Bug 1837079 - [5/10] Create WinFileDialog IPDL protocol and implementation  r=gstoll,handyman,ipc-reviewers,nika
#   
#   Create and implement a new top-level IPC protocol, `PWinFileDialog`,
#   using the primitives from bug 1833450 and the new sandboxing type from
#   the previous commit.
# ==============================================================================

diff -r 36d3afd881b4 -r d56484304ec3 ipc/glue/UtilityProcessChild.cpp
--- a/ipc/glue/UtilityProcessChild.cpp	Fri Sep 15 18:10:00 2023 +0000
+++ b/ipc/glue/UtilityProcessChild.cpp	Fri Sep 15 18:10:01 2023 +0000
@@ -5,14 +5,14 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 #include "UtilityProcessChild.h"
 
-#include "mozilla/ipc/UtilityProcessManager.h"
-#include "mozilla/ipc/UtilityProcessSandboxing.h"
+#include "mozilla/AppShutdown.h"
 #include "mozilla/dom/ContentParent.h"
 #include "mozilla/dom/JSOracleChild.h"
 #include "mozilla/dom/MemoryReportRequest.h"
 #include "mozilla/ipc/CrashReporterClient.h"
 #include "mozilla/ipc/Endpoint.h"
-#include "mozilla/AppShutdown.h"
+#include "mozilla/ipc/UtilityProcessManager.h"
+#include "mozilla/ipc/UtilityProcessSandboxing.h"
 #include "mozilla/Preferences.h"
 #include "mozilla/RemoteDecoderManagerParent.h"
 
@@ -33,6 +33,7 @@
 #if defined(XP_WIN)
 #  include "mozilla/WinDllServices.h"
 #  include "mozilla/dom/WindowsUtilsChild.h"
+#  include "mozilla/widget/filedialog/WinFileDialogChild.h"
 #endif
 
 #include "nsDebugImpl.h"
@@ -265,7 +266,7 @@
       MarkerOptions(MarkerTiming::IntervalUntilNowFrom(mChildStartTime)));
   mUtilityAudioDecoderInstance = new UtilityAudioDecoderParent();
   if (!mUtilityAudioDecoderInstance) {
-    return IPC_FAIL(this, "Failing to create UtilityAudioDecoderParent");
+    return IPC_FAIL(this, "Failed to create UtilityAudioDecoderParent");
   }
 
   mUtilityAudioDecoderInstance->Start(std::move(aEndpoint));
@@ -279,7 +280,7 @@
       MarkerOptions(MarkerTiming::IntervalUntilNowFrom(mChildStartTime)));
   mJSOracleInstance = new mozilla::dom::JSOracleChild();
   if (!mJSOracleInstance) {
-    return IPC_FAIL(this, "Failing to create JSOracleParent");
+    return IPC_FAIL(this, "Failed to create JSOracleParent");
   }
 
   mJSOracleInstance->Start(std::move(aEndpoint));
@@ -302,6 +303,29 @@
   return IPC_OK();
 }
 
+mozilla::ipc::IPCResult UtilityProcessChild::RecvStartWinFileDialogService(
+    Endpoint<widget::filedialog::PWinFileDialogChild>&& aEndpoint) {
+  PROFILER_MARKER_UNTYPED(
+      "UtilityProcessChild::RecvStartWinFileDialogService", OTHER,
+      MarkerOptions(MarkerTiming::IntervalUntilNowFrom(mChildStartTime)));
+
+  MOZ_RELEASE_ASSERT(!mFileDialogInstance,
+                     "attempted to double-start file dialog service");
+
+  auto instance = MakeRefPtr<widget::filedialog::WinFileDialogChild>();
+  if (!instance) {
+    return IPC_FAIL(this, "Failed to create WinFileDialogChild");
+  }
+
+  bool const ok = std::move(aEndpoint).Bind(instance.get());
+  if (!ok) {
+    return IPC_FAIL(this, "Failed to bind created WinFileDialogChild");
+  }
+
+  mFileDialogInstance = std::move(instance);
+  return IPC_OK();
+}
+
 mozilla::ipc::IPCResult UtilityProcessChild::RecvGetUntrustedModulesData(
     GetUntrustedModulesDataResolver&& aResolver) {
   RefPtr<DllServices> dllSvc(DllServices::Get());